File: python/nodes/count_node.py
import asyncio
import random
from typing import Dict, Any, Protocol, Union, Awaitable
from ambler import Edges, Node, Next

class Utils(Protocol):
    def log(self, message: str) -> None:
        ...
    def sleep(self, ms: int) -> Awaitable[None]:
        ...
    def random(self) -> float:
        ...

class DefaultUtils:
    def log(self, message: str) -> None:
        print(message)
    
    async def sleep(self, ms: int) -> None:
        await asyncio.sleep(ms / 1000)

    def random(self) -> float:
        return random.random()

Edge = Union[str, str] # "onCount" | "onStop"

def create(
    edges: Edges[str, str],
    utils: Utils = DefaultUtils()
) -> Node[Dict[str, Any], str]:
    async def node_fn(state: Dict[str, Any], key: str) -> Next[Dict[str, Any], str]:
        utils.log(f"Current count: {state['count']}")

        await utils.sleep(1000)

        next_state = {**state, "count": state["count"] + 1}

        stop = utils.random() > 0.7

        if stop:
            return edges["onStop"], next_state

        return edges["onCount"], next_state

    return node_fn

File: go/nodes/count_node.go
package nodes

import (
	"context"
	"fmt"
	"math/rand"
	"time"

	"github.com/argenkiwi/ambler/go"
)

// CountNodeUtils defines the side-effectful operations for CountNode.
type CountNodeUtils interface {
	Log(message string)
	Sleep(ctx context.Context, duration time.Duration) error
	Random() float64
}

// DefaultCountNodeUtils is the production implementation of CountNodeUtils.
type DefaultCountNodeUtils struct{}

func (d DefaultCountNodeUtils) Log(message string) {
	fmt.Println(message)
}

func (d DefaultCountNodeUtils) Sleep(ctx context.Context, duration time.Duration) error {
	select {
	case <-time.After(duration):
		return nil
	case <-ctx.Done():
		return ctx.Err()
	}
}

func (d DefaultCountNodeUtils) Random() float64 {
	return rand.Float64()
}

// CountNodeEdges defines the possible transitions from CountNode.
type CountNodeEdges[K comparable] struct {
	OnCount *K
	OnStop  *K
}

// CreateCountNode creates a new CountNode.
func CreateCountNode[S HasCount[S], K comparable](edges CountNodeEdges[K], utils CountNodeUtils) ambler.Node[S, K] {
	if utils == nil {
		utils = DefaultCountNodeUtils{}
	}

	return func(ctx context.Context, state S, key K) (ambler.Next[S, K], error) {
		utils.Log(fmt.Sprintf("Current count: %d", state.GetCount()))

		err := utils.Sleep(ctx, 1*time.Second)
		if err != nil {
			return ambler.Next[S, K]{}, err
		}

		nextState := state.WithCount(state.GetCount() + 1)

		stop := utils.Random() > 0.7

		if stop {
			return ambler.Next[S, K]{NodeID: edges.OnStop, State: nextState}, nil
		}

		return ambler.Next[S, K]{NodeID: edges.OnCount, State: nextState}, nil
	}
}
